crates/control/src/services/discovery.rs

```diff
@@ -0,0 +1,190 @@
+use models::Schema;
+use models::ShardTemplate;
+
+use crate::models::connector_images::ConnectorImage;
+use crate::models::connectors::Connector;
+use crate::services::connectors::DiscoveryOptions;
+
+#[derive(Debug, Deserialize)]
+pub struct DiscoverResponse {
+    pub bindings: Vec<DiscoveredBinding>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct DiscoveredBinding {
+    /// A recommended display name for this discovered binding.
+    pub recommended_name: String,
+    /// JSON-encoded object which specifies the endpoint resource to be captured.
+    #[serde(rename = "resourceSpec")]
+    pub resource_spec_json: Object,
+    /// JSON schema of documents produced by this binding.
+    #[serde(rename = "documentSchema")]
+    pub document_schema_json: Object,
+    /// Composite key of documents (if known), as JSON-Pointers.
+    pub key_ptrs: Vec<String>,
+}
+
+impl DiscoveredBinding {
+    fn key(&self) -> models::CompositeKey {
+        CompositeKey::new(
+            self.key_ptrs
+                .iter()
+                .map(JsonPointer::new)
+                .collect::<Vec<JsonPointer>>(),
+        )
+    }
+
+    pub fn schema_url(&self) -> Schema {
+        Schema::Url(RelativeUrl::new(self.schema_name()))
+    }
+
+    pub fn schema_name(&self) -> String {
+        format!("{}.schema.json", self.recommended_name)
+    }
+}
+
+pub struct DiscoveredCatalog {
+    connector: Connector,
+    image: ConnectorImage,
+    config: Object,
+    bindings: Vec<DiscoveredBinding>,
+    options: DiscoveryOptions,
+}
+
+impl DiscoveredCatalog {
+    pub fn new(
+        connector: Connector,
+        image: ConnectorImage,
+        config: Object,
+        bindings: Vec<DiscoveredBinding>,
+        options: DiscoveryOptions,
+    ) -> Self {
+        Self {
+            connector,
+            image,
+            config,
+            bindings,
+            options,
+        }
+    }
+
+    pub fn image(&self) -> &ConnectorImage {
+        &self.image
+    }
+
+    pub fn render_catalog(&self) -> Catalog {
+        Catalog {
+            captures: self.capture_definitions(),
+            collections: self.discovered_collections(),
+            ..Default::default()
+        }
+    }
+
+    pub fn render_config(&self) -> Config {
+        Config::Inline(self.config.clone())
+    }
+
+    pub fn render_schemas(&self) -> BTreeMap<String, Schema> {
+        let mut schemas = BTreeMap::new();
+        for binding in self.bindings.iter() {
+            schemas.insert(
+                binding.schema_name(),
+                Schema::Object(binding.document_schema_json.clone()),
+            );
+        }
+        schemas
+    }
+
+    pub fn name(&self) -> String {
+        format!("{}.flow.json", self.connector.codename())
+    }
+
+    pub fn config_name(&self) -> String {
+        format!("{}.config.json", self.connector.codename())
+    }
+
+    fn capture_definitions(&self) -> BTreeMap<Capture, CaptureDef> {
```

Comment:
Are there cases where we could discover more than one capture? I was surprised to see this being a BTreeMap, and I'm not sure if this is just a bit of future proofing, or if I'm just missing something.